### ISSUE TITLE
Why casting to uint 64 when being stored in uint 32?

### ISSUE THREAD

**Issue by meepingsnesroms** [joncampbell123/dosbox-x#249]
`void CPU_CMPXCHG8B(PhysPt eaa) {

    uint32_t hi,lo;
    /* NTS: We assume that, if reading doesn't cause a page fault, writing won't either */
    hi = (uint64_t)mem_readd(eaa+(PhysPt)4);
    lo = (uint64_t)mem_readd(eaa);

    LOG_MSG("Experimental CMPXCHG8B implementation executed. EDX:EAX=0x%08lx%08lx ECX:EBX=0x%08lx%08lx EA=0x%08lx MEM64=0x%08lx%08lx",
        (unsigned long)reg_edx,
        (unsigned long)reg_eax,
        (unsigned long)reg_ecx,
        (unsigned long)reg_ebx,
        (unsigned long)eaa,
        (unsigned long)hi,
        (unsigned long)lo);

    /* Compare EDX:EAX with 64-bit DWORD at memaddr 'eaa'.
     * if they match, ZF=1 and write ECX:EBX to memaddr 'eaa'.
     * else, ZF=0 and load memaddr 'eaa' into EDX:EAX */
    if (reg_edx == hi && reg_eax == lo) {
        mem_writed(eaa+(PhysPt)4,reg_ecx);
        mem_writed(eaa,          reg_ebx);
		SETFLAGBIT(ZF,true);
    }
    else {
		SETFLAGBIT(ZF,false);
        reg_edx = hi;
        reg_eax = lo;
    }
}`

**Comment by joncampbell123** [comment]
Good point. My mistake. Not harmful to the code though.

**Comment by meepingsnesroms** [comment]
I noticed this when porting the new opcode to retroarch, do you have any interest in helping port parts of your project to retroarch?

Mainly Voodoo Gfx.

**Comment by joncampbell123** [comment]
Not sure how well I can port the 3Dfx emulation, since I didn't write it. It was inherited from Daum's fork. But I can help with the code I have written, like the new mixer, video capture, floppy and IDE emulation.

**Comment by meepingsnesroms** [comment]
Floppy,ide and mixer would help, are they significantly faster of more accurate?
The video capture is part of retroarch already and dumps the raw output buffers so it is 100% accurate with no extra code on the emulators side.

New opcodes and hardware 8086 mode would be the best currently.

The URL is here and I have write access so when you make a pull request it will likely be merged within an hour.

[https://github.com/libretro/dosbox-libretro](url)

**Comment by joncampbell123** [comment]
Floppy and IDE emulation is written to be as accurate as I can make it. The new mixer code fixes an issue that bothers me with DOSBox where DOSBox emulation time is driven by SDL ticks but audio rendering is driven by the sound card. They don't necessarily line up, and until I fixed it I had problems with audio and video slowly drifting out of sync. The new mixer code also offers lowpass filtering and slew emulation to better emulate what the old sound cards (especially Sound Blaster) actually sound like.

I have cloned your repository, to develop on it when I have time.
